### DISCUSSION TITLE
Filtering two different model fields

### DISCUSSION THREAD

**Discussion by pljspahn** [carltongibson/django-filter#1427] [Q&A] [answered]
I am modeling products (plants in my case) that have a foreign key relationship to a genus model.

```
class Product(models.Model):
    sku = models.CharField('SKU', max_length=14)
    name = models.CharField('Name', max_length=32)
    price = models.DecimalField('Price', max_digits=8, decimal_places=2)
    qoh = models.IntegerField('Quantity on Hand', default=0)
    genus = models.ForeignKey(Genus, on_delete=models.SET_NULL, null=True, blank=True)
    
    class Meta:
        ordering = ['sku']

    def __str__(self):
        return self.name

class Genus(models.Model):
    #ie Apple
    common = models.CharField('Common Genus', max_length=100)
    #ie Malus
    latin = models.CharField('Latin Genus', max_length=100)

    class Meta:
        ordering = ['common']
        verbose_name_plural = 'genera'

    def __str__(self):
        return self.common

```

I have a `ModelMultipleChoiceFilter` set up to filter by common genus. This works just like I'd expect.

```
class ProductFilter(django_filters.FilterSet):
    genus__common = django_filters.ModelMultipleChoiceFilter(queryset=Genus.objects.all(), label='Genus')
    class Meta:
        model = Product
        fields = ['genus__common']
```
I added a second filter to allow filtering by Latin name:

```
class LatinFilter(django_filters.FilterSet):
    genus__latin = django_filters.ModelMultipleChoiceFilter(queryset=Genus.objects.all(), label='Latin Genus')
    class Meta:
        model = Product
        fields = ['genus__latin']
```
This is used in my template like:
```
<form id="filters" method="GET">
    {{filter.form}}
    {{latin_filter.form}}
    <input type="submit">
</form>
```
This approach doesn't work as:
1. The form uses `__str__` method of the Genus model to display the common genus instead of the Latin genus
2. It just doesn't filter correctly, returning all objects
3. Optionally, I'd like to change the ordering for `LatinFilter` so that they order by latin instead of common.

Where am I going wrong here?

**Comment by rpkilby** [comment, marked as answer]
> The form uses `__str__` method of the Genus model to display the common genus instead of the Latin genus

By default, choice labels are generated using `__str__`. To change this, you'll need to implement `label_from_instance` as described at the very bottom of the [`ModelChoiceField`](https://docs.djangoproject.com/en/3.0/ref/forms/fields/#modelchoicefield) docs. Then, you'll need to provide the `field_class` to your `ModelMultipleChoiceFilter`. 

> It just doesn't filter correctly, returning all objects

This will still depend on your view, but I can at least say that you don't need to have two separate filterset classes. `genus__common` and `genus__latin` can be members of the same filterset.

> Optionally, I'd like to change the ordering for LatinFilter so that they order by latin instead of common.

You've provide the `queryset` to the filter, so you tack on an `.order_by()` call there.

**Comment by pljspahn** [comment]
Thanks so much. I guess I was not fully understanding how `__` works and thought there was some extra magic happening there that wasn't.

I changed my `ProductFilter` to the following and added `LatinChoiceField` and `LatinFilter`:

```
class LatinChoiceField(ModelMultipleChoiceField):
    def label_from_instance(self, obj):
        return obj.latin

class LatinFilter(django_filters.ModelMultipleChoiceFilter):
    field_class = LatinChoiceField

class ProductFilter(django_filters.FilterSet):
    genus__common = django_filters.ModelMultipleChoiceFilter(queryset=Genus.objects.all(), label='Genus')
    genus__latin = LatinFilter(queryset=Genus.objects.all().order_by('latin'), label='Latin', field_name='genus')
    class Meta:
        model = Product
        fields = ['genus__common', 'genus__latin']
 ```
I still don't really understand why the `genus__latin` Filter doesn't work unless I add `field_name='genus'` - that just doesn't make sense to me since I would expect the value to be `field_name='genus__latin'`, but for whatever reason it does work.
